Griffon master stream7 feature bjs2 40436

UrmatBolotbek opened the pull request.

UrmatBolotbek committed:
add SkillOfferDto and RecommendationDto

UrmatBolotbek committed:
add DataValidationException and ErrorMessage

UrmatBolotbek committed:
change implemented interface CrudRepository to JpaRepository in DataValidationException and ErrorMessage

UrmatBolotbek committed:
add RecommendationController

UrmatBolotbek committed:
update ErrorMessage, RecommendationDto and SkillOfferDto

UrmatBolotbek committed:
update RecommendationController

UrmatBolotbek committed:
update ErrorMessage and SkillOfferDto

UrmatBolotbek committed:
add RecommendationMapper and SkillOfferMapper

UrmatBolotbek committed:
add RecommendationService and update RecommendationController

UrmatBolotbek committed:
add method updateRecommendation into RecommendationController

UrmatBolotbek committed:
add method update into RecommendationService

UrmatBolotbek committed:
add method delete into RecommendationService

UrmatBolotbek committed:
add method deleteRecommendation into RecommendationController

UrmatBolotbek committed:
add method toDtoList into RecommendationMapper

UrmatBolotbek committed:
add method getAllUserRecommendations into RecommendationService

UrmatBolotbek committed:
add method getAllUserRecommendations into RecommendationController

UrmatBolotbek committed:
add method getAllGivenRecommendations into RecommendationController

UrmatBolotbek committed:
add method getAllGivenRecommendations into RecommendationService

RomaNifontov commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,65 @@
+        if (recommendationDto.getContent() == null || recommendationDto.getContent().isBlank()) {
+            throw new DataValidationException(ErrorMessage.RECOMMENDATION_CONTENT);
+        }
+        if (recommendationDto.getAuthorId() == null) {
```
Валидация в задании нужна, чтобы проверить на наличие текста, думаю мы заранее должны будем в другом методе проверить на то, что валидации присваивается айди автора и айди рекомендации.

UrmatBolotbek replied:
несомненно, но я дополнительно решил чтобы в контроллере был только один метод валидации и проверял он все важные поля DTO. Это решении было согласовано с Костей.

RomaNifontov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,169 @@
+        recommendationDto.setId(recommendationId);
+        addSkillOffersAndGuarantee(recommendationDto);
+
+        Recommendation recommendation = recommendationRepository.findById(recommendationDto.getId())
```
Думаю лучше вынести проверку на наличие рекомендации в отдельный метод

7Alena7 replied:
Здесь лучше с маппером. После валидации будет что-то вроде 
Recommendation recommendation = recommendationMapper.toEntity(reccomendationDto);
recomendation = recomendationRepository.save(recomendation);
 return recomendationMapper.toDto(recomendation);

UrmatBolotbek replied:
сделано

UrmatBolotbek replied:
например в методе 
 @Transactional
    public RecommendationDto create(RecommendationDto recommendationDto) {
        checkIfOfferedSkillsExist(recommendationDto); 
после этой строки метод закончится свое выполнение при описанных мною выше условиях
        checkIfAcceptableTimeForRecommendation(recommendationDto);

        Long recommendationId = recommendationRepository.create(
                recommendationDto.getAuthorId(),
                recommendationDto.getReceiverId(),
                recommendationDto.getContent());

        recommendationDto.setId(recommendationId);
        addSkillOffersAndGuarantee(recommendationDto);

        return recommendationMapper.toDto(loadUpdatedRecommendation(recommendationDto.getId()));
    }

RomaNifontov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,169 @@
+    private void checkIfOfferedSkillsExist(RecommendationDto recommendationDto) {
+        List<SkillOfferDto> skillOfferDtoList = recommendationDto.getSkillOffers();
+        if (skillOfferDtoList == null || skillOfferDtoList.isEmpty()) {
+            return;
```
Как я понял мы проверяем, что лист либо ноль либо пустой и просто возвращаемся. Тут скорее всего должно быть брошено исключение, что навыков в рекомендаций вообще нет, нам нельзя выполнять дальше метод create. 
"Предложенные скиллы в рамках рекомендации не могут дублироваться и должны быть существующими скиллами в системе. ", а если нет скиллов, то почему мы просто выходим из валидации, а не кидаем исключение

UrmatBolotbek replied:
как Влад не однократно повторяет в своих видео, не нужно всегда выкидывать исключения или ограничивать действия пользователя и тыкать его лицом в ошибки. Данная форма записи как раз таки про это. Если условие выполняется то return просто завершает работу этого метода. А в методах которых используется данный метод, в случае когда условие if (skillOfferDtoList == null || skillOfferDtoList.isEmpty()) выполняется, то код последующий за ним попросту не будет выполняться

UrmatBolotbek replied:
например в методе
@transactional
public RecommendationDto create(RecommendationDto recommendationDto) {
checkIfOfferedSkillsExist(recommendationDto);
после этой строки метод закончится свое выполнение при описанных мною выше условиях
checkIfAcceptableTimeForRecommendation(recommendationDto);

    Long recommendationId = recommendationRepository.create(
            recommendationDto.getAuthorId(),
            recommendationDto.getReceiverId(),
            recommendationDto.getContent());

    recommendationDto.setId(recommendationId);
    addSkillOffersAndGuarantee(recommendationDto);

    return recommendationMapper.toDto(loadUpdatedRecommendation(recommendationDto.getId()));
}

UrmatBolotbek replied:
например в этом методе после строки checkIfOfferedSkillsExist(recommendationDto);
 метод закончится свое выполнение при описанных мною выше условиях

@transactional
public RecommendationDto create(RecommendationDto recommendationDto) {
checkIfOfferedSkillsExist(recommendationDto);
checkIfAcceptableTimeForRecommendation(recommendationDto);

    Long recommendationId = recommendationRepository.create(
            recommendationDto.getAuthorId(),
            recommendationDto.getReceiverId(),
            recommendationDto.getContent());

    recommendationDto.setId(recommendationId);
    addSkillOffersAndGuarantee(recommendationDto);

    return recommendationMapper.toDto(loadUpdatedRecommendation(recommendationDto.getId()));
}

UrmatBolotbek replied:
например в этом методе после строки checkIfOfferedSkillsExist(recommendationDto);
метод закончится свое выполнение при описанных мною выше условиях

@transactional
public RecommendationDto create(RecommendationDto recommendationDto) {
checkIfOfferedSkillsExist(recommendationDto);
checkIfAcceptableTimeForRecommendation(recommendationDto);
Long recommendationId = recommendationRepository.create(
        recommendationDto.getAuthorId(),
        recommendationDto.getReceiverId(),
        recommendationDto.getContent());
recommendationDto.setId(recommendationId);
addSkillOffersAndGuarantee(recommendationDto);
return recommendationMapper.toDto(loadUpdatedRecommendation(recommendationDto.getId()));
}

7Alena7 replied:
Мне кажется все таки нужно кидать ошибку. Если не кинуть ту то в подобном случае из LinkedIn летели бы сообщения что Пользователь endorsed ваш навык. А что за навык непонятно, он пустой. Короче, я голосую за exception

RomaNifontov replied:
Так может я что-то не понимаю, но почему у тебя метод закончит работу если ты return вообще в другом методе сделал? 
![image](https://github.com/user-attachments/assets/d95b8d7d-16b6-421c-b643-97864de714c2)

RomaNifontov replied:
Я вот так увидел твой код, может я ошибаюсь, поправь если что

RomaNifontov replied:
если ты не хочешь кидать исключения то check методы могут возвращать boolean и там уже по условию дальше

UrmatBolotbek replied:
Да, запись if (skillOfferDtoList == null || skillOfferDtoList.isEmpty()) { return; } в методе checkIfOfferedSkillsExist является допустимой и обычной практикой в программировании. Она используется для оптимизации кода и предотвращения лишних операций, когда нет необходимости выполнять проверки или другие действия, если условие не выполняется.

В контексте метода checkIfOfferedSkillsExist, если список skillOfferDtoList равен null (что может произойти, если recommendationDto или его список skillOffers не инициализированы) или если список пустой (то есть в recommendationDto нет предложений по навыкам), то нет смысла выполнять дальнейшие проверки на существование навыков. Поэтому метод просто возвращает управление, не выбрасывая исключение и не выполняя дополнительных проверок.

Такой подход улучшает читаемость кода и помогает избежать лишних операций, что важно для обеспечения эффективной работы программы. Если условие if (skillOfferDtoList == null || skillOfferDtoList.isEmpty()) выполнилось (то есть список skillOfferDtoList пуст или равен null), то метод checkIfOfferedSkillsExist(recommendationDto) завершится оператором return;. В этом случае код, который следует за этим условием, не будет выполнен.

Таким образом, если список предложений по навыкам (skillOffers) в recommendationDto пуст или не инициализирован:

Выполнится оператор return;, и выполнение метода create будет завершено.
Не будет выполнено создание новой рекомендации в базе данных с помощью recommendationRepository.create(...).
Не будет установлен идентификатор (id) для recommendationDto.
Не будет выполнено добавление предложений по навыкам и гарантий с помощью addSkillOffersAndGuarantee(recommendationDto).
В итоге, метод вернет результат работы recommendationMapper.toDto(loadUpdatedRecommendation(recommendationDto.getId())), но loadUpdatedRecommendation будет вызван с null в качестве id, что может привести к исключению 
а в этом методе уже есть обработка исключения, т.е. не надо выкидывать везде и обрабатывать везде
private Recommendation getRecommendation(Long recommendationId) {
        return recommendationRepository.findById(recommendationId)
                .orElseThrow(() -> {
                    log.error("Recommendation with id {} not found", recommendationId);
                    return new NoSuchElementException(
                            String.format("There is no recommendation with id = %d", recommendationId));
                });
    }

RomaNifontov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,169 @@
+        skillOfferRepository.deleteAllByRecommendationId(recommendationDto.getId());
+        addSkillOffersAndGuarantee(recommendationDto);
+
+        Recommendation updatedRecommendation = recommendationRepository.findById(recommendationDto.getId())
```
Вынести отдельно метод валидации

UrmatBolotbek committed:
update method updateRecommendation into RecommendationService and RecommendationController

RomaNifontov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Transactional
+    public void delete(long id) {
```
проверка валидации

7Alena7 commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,19 @@
+@Mapper(componentModel = "spring", uses = SkillOfferMapper.class)
+public interface RecommendationMapper {
+
+    @Mapping(source = "author.id", target = "authorId")
```
Mapper лезет не в саму базу данных, а сравнивает entity и dto. Поэтому тут не нужно в ручную мапить поля, они и так одинаково называются

UrmatBolotbek replied:
здесь согласен

7Alena7 commented on src/main/java/school/faang/user_service/mapper/SkillOfferMapper.java:
```diff
@@ -0,0 +1,14 @@
+@Mapper(componentModel = "spring")
+public interface SkillOfferMapper {
+
+    @Mapping(source = "skill.id", target = "skillId")
```
тут тоже не нужно мапить поля

UrmatBolotbek replied:
а вот тут он берет поля из класса Skill, а не SkillOffer поэтому нужно указать поля

7Alena7 reviewed (commented):
Вроде все хорошо кроме использования mapstruct. Посмотри вот здесь последнее видео https://faang-school.com/courses/4jnzmndg/vwze26kd и еще по деталям можно быстренько спросить у ChatGpt  и потом перепроверть его вот тут https://mapstruct.org/documentation/stable/reference/html/#updating-bean-instances

UrmatBolotbek committed:
обновил мапперы

UrmatBolotbek committed:
добавил логи в сервис
вынес логику валидации RecommendationDto в отдельный класс
поправил иерархию папок

7Alena7 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,152 @@
+
+        recommendationDtoValidator.validateRecommendation(recommendationDto);
+
+        Long recommendationId = recommendationRepository.create(
```
Вот тут проще маппером тоже будет

UrmatBolotbek replied:
у меня по заданию стоит для создания в бд новой записи рекомендации использовать данный метод.

Liberate520 replied:
Алена говорит, что надо превратить дто в entity, а затем дто отправить в метод create. Это верное замечание

UrmatBolotbek replied:
но у меня метод принимает не entity а лишь некоторые поля из дто

Liberate520 replied:
Ааа... там кто то написал нативный метод... метод create это стандартный метод JpaRepository или CrudRepository. В него должна передаваться сущность с которой работает репозиторий. В данном случае объект Recommendation. Для этого тебе надо превратить свое дто в этот объект. По сути в таблице recommendation нет обязательных полей, кроме id, который создается автоматически, поэтому можно передать любой набор данных и кто то решил, что ему надо передать именно три поля и написал нативный метод. Но на самом деле эти же три поля будут сохранены, если в дто и было только три поля. В общем сколько полей будет, столько и будет сохранено. Правильно в методе create конечно же сохранить все поля, которые тебе пришли в дто

Liberate520 replied:
пользуйся методом save, а не create

Liberate520 replied:
метод save вернет объект Recommendation, в котором будет например id, которое было сгенерировано бд для этой записи

7Alena7 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,152 @@
+        recommendationDto.setId(id);
+        recommendationDtoValidator.validateRecommendation(recommendationDto);
+
+        recommendationRepository.update(
```
и вот тут

Liberate520 replied:
Тут тоже будет save

RomaNifontov commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private void addGuaranteeToSkill(RecommendationDto recommendationDto, Skill skill) {
+        User receiver = userRepository.findById(recommendationDto.getReceiverId())
```
Добавь два метода валидации в отдельную свою папку по валидации)

UrmatBolotbek committed:
подправил одну строку сообщения

UrmatBolotbek committed:
разделил DTO на DTO запроса и ответа и обновил места их использования в контроллере и сервисе

UrmatBolotbek committed:
добавил логи

IsmoilovMunir commented on src/main/java/school/faang/user_service/dto/recommendation/RecommendationDto.java:
```diff
@@ -0,0 +1,16 @@
+package school.faang.user_service.dto.recommendation;
+
+import lombok.Data;
+
```
отступ лишнее

UrmatBolotbek replied:
этот отступ идея сама оставляет когда ты правишь код

IsmoilovMunir commented on src/main/java/school/faang/user_service/mapper/recommendation/RecommendationMapper.java:
```diff
@@ -0,0 +1,20 @@
+
+@Mapper(componentModel = "spring", unmappedSourcePolicy = ReportingPolicy.IGNORE, uses = SkillOfferMapper.class)
+public interface RecommendationMapper {
+
```
почему ты mapper только по dto делал? В дальнейшим mapper нам по entity тоже нужно будет

UrmatBolotbek replied:
у меня по заданию нужно использовать метод из репозитория который сам создает объект и сохраняет его в бд. этот момент уже улажен

IsmoilovMunir commented on src/main/java/school/faang/user_service/mapper/recommendation/SkillOfferMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Mapping(source = "skill.id", target = "skillId")
+    @Mapping(source = "skill.title", target = "skillTitle")
+    SkillOfferDto toDto(SkillOffer skillOffer);
```
тут тоже Mapper toEntity нет

UrmatBolotbek replied:
этот момоент тоже в тз прописан и улажен

Liberate520 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,65 @@
+        return recommendationService.update(id, updatedRecommendationDto);
+    }
+
+    @DeleteMapping("/delete/{id}")
```
Это неверно, но это на следующей неделе расскажут)

Liberate520 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,152 @@
+        List<SkillOfferDto> skillOfferDtoList = recommendationDto.getSkillOffers();
+        if (skillOfferDtoList == null || skillOfferDtoList.isEmpty()) {
+            return;
+        }
```
Я видел этот комментарий от ребят. Ты делаешь return когда данные не валидны, а значит 69 и далее строчки будут выполнены. А так быть не должно

UrmatBolotbek committed:
изменил мапперы чтобы тесты работали корректно

UrmatBolotbek committed:
добавил новое сообщение

UrmatBolotbek committed:
добавил выброс исключения в метод addSkillOffersAndGuarantee

UrmatBolotbek committed:
добавил аннотацию @Builder

UrmatBolotbek committed:
написал тесты на метод create сервиса RecommendationService

Liberate520 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,157 @@
+    private void addSkillOffersAndGuarantee(RequestRecommendationDto requestRecommendationDto) {
+        List<RequestSkillOfferDto> requestSkillOfferDtoList = requestRecommendationDto.getSkillOffers();
+        if (requestSkillOfferDtoList == null || requestSkillOfferDtoList.isEmpty()) {
+            log.debug("No skill offers to process for recommendation with id {}", requestRecommendationDto.getId());
```
это будет warn

Liberate520 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,157 @@
+    private Recommendation getRecommendation(Long recommendationId) {
+        return recommendationRepository.findById(recommendationId)
+                .orElseThrow(() -> {
+                    log.error("Recommendation with id {} not found", recommendationId);
```
тоже будет warn

Liberate520 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        for (RequestSkillOfferDto requestSkillOfferDto : requestSkillOfferDtoList) {
+            skillOfferRepository.create(requestSkillOfferDto.getSkillId(), requestRecommendationDto.getId());
```
Тоже поменять на метод save

UrmatBolotbek committed:
написал тесты на метод create сервиса RecommendationService

UrmatBolotbek committed:
дописал тесты на сервис и валидатор

UrmatBolotbek committed:
сделал выброс исключения в валидаторе и покрыл тестами

Liberate520 commented on src/main/java/school/faang/user_service/dto/recommendation/RequestRecommendationDto.java:
```diff
@@ -0,0 +1,18 @@
+    private Long receiverId;
+    private String content;
+    private List<RequestSkillOfferDto> skillOffers;
+    private LocalDateTime createdAt;
```
В запросе не надо указывать дату создания. Бд создаст это значение автоматически

Liberate520 commented on src/main/java/school/faang/user_service/dto/recommendation/RequestSkillOfferDto.java:
```diff
@@ -0,0 +1,13 @@
+
+@Data
+@Builder
+public class RequestSkillOfferDto {
```
Этот же объект у тебя не в запросе и ответе участвует. Это просто объект для группировки id. Можно было одним обойтись

Liberate520 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationService.java:
```diff
@@ -0,0 +1,159 @@
+
+        Recommendation recommendation = getRecommendation(requestRecommendationDto.getId());
+
+        for (RequestSkillOfferDto requestSkillOfferDto : requestSkillOfferDtoList) {
+            Skill skill = getSkill(requestSkillOfferDto.getSkillId());
+            SkillOffer skillOffer = SkillOffer.builder()
+                    .skill(skill)
+                    .recommendation(recommendation)
+                    .build();
+
+            skillOfferRepository.save(skillOffer);
```
Я же тебе вроде бы объяснял, что когда ты сохраняешь объект Recommendation, то ты автоматически сохраняешь все его содержимое, в том числе список skillOffers. Нет необходимости все это сохранять повторно.

Liberate520 replied:
Тебе надо перед сохранением рекомендации создать записи о гарантах и потом сохранить объект рекомендации

Liberate520 commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationServiceTest.java:
```diff
@@ -0,0 +1,226 @@
+        requestRecommendationDto.setSkillOffers(null);
+        when(recommendationMapper.toEntity(requestRecommendationDto)).thenReturn(recommendation);
+        when(recommendationRepository.save(any(Recommendation.class))).thenReturn(recommendation);
+
```
Проверить, что метод save в этом методе не вызывался. Для этого используй verify

UrmatBolotbek committed:
Исправил момент дублирования списка SkillOffers в методах создания и обновления рекомендаций.

Liberate520 commented on src/main/java/school/faang/user_service/exception/DataValidationException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.exception;
```
Добавь пакет recommendation

UrmatBolotbek committed:
давабил пакет recommendation в папку exception
